File: src/Services/LeagueSportRules.cs
namespace Sportarr.Api.Services;

/// <summary>
/// Single source of truth for "which sports have no home/away team structure".
/// These leagues auto-monitor on add (no team selection required) and bypass
/// team-based event filtering. Must stay in sync with the frontend helper
/// isTeamlessSport in frontend/src/utils/leagueSportRules.ts.
/// </summary>
public static class LeagueSportRules
{
    private static readonly string[] TeamlessSports = new[]
    {
        "Fighting", "Cycling", "Motorsport", "Golf", "Darts",
        "Climbing", "Gambling", "Badminton", "Table Tennis", "Snooker"
    };

    /// <summary>
    /// Returns true for sports/leagues that do not have meaningful home/away
    /// teams. Individual tennis tours (ATP/WTA) also qualify, but team-based
    /// tennis competitions (Fed Cup, Davis Cup, Olympics, Billie Jean King Cup)
    /// do not.
    /// </summary>
    public static bool IsTeamlessSport(string? sport, string? leagueName)
    {
        if (string.IsNullOrEmpty(sport)) return false;
        if (TeamlessSports.Contains(sport, System.StringComparer.OrdinalIgnoreCase)) return true;
        return IsIndividualTennisLeague(sport, leagueName ?? string.Empty);
    }

    public static bool IsIndividualTennisLeague(string sport, string leagueName)
    {
        if (!sport.Equals("Tennis", System.StringComparison.OrdinalIgnoreCase)) return false;
        var nameLower = leagueName.ToLowerInvariant();
        var teamBased = new[] { "fed cup", "davis cup", "olympic", "billie jean king" };
        if (teamBased.Any(t => nameLower.Contains(t))) return false;
        var individualTours = new[] { "atp", "wta" };
        return individualTours.Any(t => nameLower.Contains(t));
    }
}

File: frontend/src/utils/leagueSportRules.ts
// League sport classification helpers.
//
// These rules centralize the logic for "does this sport behave like a
// team-based league?" so LeagueSearchPage and LeagueDetailPage stay aligned.

/**
 * Returns true for motorsports (F1, NASCAR, WRC, ...).
 * Motorsports have no meaningful home/away team - all participants race
 * in every event, so the league is always considered monitored.
 */
export function isMotorsport(sport: string): boolean {
  const motorsports = [
    'Motorsport', 'Racing', 'Formula 1', 'F1', 'NASCAR', 'IndyCar',
    'MotoGP', 'WEC', 'Formula E', 'Rally', 'WRC', 'DTM', 'Super GT',
    'IMSA', 'V8 Supercars', 'Supercars', 'Le Mans',
  ];
  return motorsports.some(s => sport.toLowerCase().includes(s.toLowerCase()));
}

/**
 * Returns true for golf.
 * Golf tournaments have all players competing together - no home/away teams.
 */
export function isGolf(sport: string): boolean {
  return sport.toLowerCase() === 'golf';
}

/**
 * Returns true for individual-format tennis leagues (ATP/WTA tours).
 * Individual tours have no meaningful team data - all events should sync.
 * Team-based competitions (Fed Cup, Davis Cup, Olympics) return false.
 */
export function isIndividualTennis(sport: string, leagueName: string): boolean {
  if (sport.toLowerCase() !== 'tennis') return false;
  const nameLower = leagueName.toLowerCase();
  const individualTours = ['atp', 'wta'];
  const teamBased = ['fed cup', 'davis cup', 'olympic', 'billie jean king'];
  if (teamBased.some(t => nameLower.includes(t))) return false;
  return individualTours.some(t => nameLower.includes(t));
}

/**
 * Returns true for fighting sports (UFC, Boxing, MMA, Wrestling, etc.).
 * Fighting events use multi-part structure (Early Prelims, Prelims, Main Card).
 */
export function isFightingSport(sport: string): boolean {
  const fightingSports = ['Fighting', 'MMA', 'UFC', 'Boxing', 'Kickboxing', 'Wrestling'];
  return fightingSports.some(s => sport.toLowerCase().includes(s.toLowerCase()));
}

/**
 * Returns true for individual-player racket/cue sports (Badminton, Table Tennis, Snooker).
 * Tournaments feature individual players, not teams - all events should sync without team filtering.
 */
export function isIndividualRacketOrCueSport(sport: string): boolean {
  const s = sport.toLowerCase();
  return s === 'badminton' || s === 'table tennis' || s === 'snooker';
}

/**
 * Darts matches are between individual players, not teams.
 */
export function isDarts(sport: string): boolean {
  return sport.toLowerCase() === 'darts';
}

/**
 * Climbing competitions are individual climbers, not teams.
 */
export function isClimbing(sport: string): boolean {
  return sport.toLowerCase() === 'climbing';
}

/**
 * Gambling (Poker, WSOP) events are individual players in tournaments, not teams.
 */
export function isGambling(sport: string): boolean {
  return sport.toLowerCase() === 'gambling';
}

/**
 * Returns true for sports/leagues that have no meaningful home/away team structure.
 * These leagues should auto-monitor on add (no team selection required) and must
 * stay in sync with the backend's sportsWithoutTeamFiltering list in
 * LeagueEventSyncService.cs.
 */
export function isTeamlessSport(sport: string, leagueName: string): boolean {
  return (
    isMotorsport(sport) ||
    isGolf(sport) ||
    isDarts(sport) ||
    isClimbing(sport) ||
    isGambling(sport) ||
    isIndividualRacketOrCueSport(sport) ||
    isIndividualTennis(sport, leagueName)
  );
}
